TINKERPOP-1085 Added gremlin-archetypes and related submodules

spmallette opened the pull request:
https://issues.apache.org/jira/browse/TINKERPOP-1085

Established the model for "examples" in TinkerPop using Maven archetypes. Provided two archetypes: gremlin-archetype-server and gremlin-archetype-tinkergraph.

Note that each archetype has tests to validate their operations and the tests are configured to execute on standard execution of `mvn clean install`.  You can manually try out the achetypes with a command like this:

```text
$ mvn archetype:generate -DarchetypeGroupId=org.apache.tinkerpop -DarchetypeArtifactId=gremlin-archetype-server -DarchetypeVersion=3.1.2-SNAPSHOT -DgroupId=com.my -DartifactId=app
```

which will create a project (based on the `gremlin-archetype-server` module) in a directory called "app".  

VOTE +1

okram commented:
VOTE +1. When you merge can you publish fresh docs to 3.1.2-SNAPSHOT so I can get a better sense of the organization of things?

pluradj commented on CHANGELOG.asciidoc:
```diff
@@ -26,6 +26,9 @@ image::https://raw.githubusercontent.com/apache/incubator-tinkerpop/master/docs/
 TinkerPop 3.1.2 (NOT OFFICIALLY RELEASED YET)
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+* Added the `gremlin-archetype-driver` archetype that demonstrates
```
no `gremlin-archetype-driver`. this line can be removed.

pluradj commented on docs/src/reference/gremlin-applications.asciidoc:
```diff
@@ -1880,3 +1880,25 @@ The following shows the output for `HadoopGraph`:
+
+[source,shell]
+$ mvn archetype:generate -DarchetypeGroupId=org.apache.tinkerpop -DarchetypeArtifactId=gremlin-archetype-server
+      -DarchetypeVersion=3.1.2-incubating -DgroupId=com.my -DartifactId=app
```
Suggest adding `-Dversion=0.1 -DinteractiveMode=false`

pluradj commented on gremlin-archetype/gremlin-archetype-tinkergraph/src/main/resources/archetype-resources/README.asciidoc:
```diff
@@ -0,0 +1,35 @@
+--------------------
+
+[source,text]
+mvn clean package
```
need to make sure that the Gremlin Server is not running, otherwise the unit test fails.

pluradj commented on gremlin-archetype/gremlin-archetype-server/src/main/resources/archetype-resources/README.asciidoc:
```diff
@@ -0,0 +1,67 @@
+Build and run this project as follows:
+
+[source,text]
+mvn clean package
```
need to make sure that the Gremlin Server is not running, otherwise the unit test fails.

pluradj commented on gremlin-archetype/gremlin-archetype-server/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,80 @@
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-log4j12</artifactId>
+            <version>${slf4j.version}</version>
+            <scope>test</scope>
```
suggest making a compile dependency and adding `log4j` also to avoid the `log4j:WARN No appenders could be found for logger`

pluradj commented on gremlin-archetype/gremlin-archetype-server/src/main/resources/archetype-resources/README.asciidoc:
```diff
@@ -0,0 +1,67 @@
+
+[source,text]
+mvn clean package
+mvn exec:java -Dexec.mainClass="${package}.App"
```
if you add the `log4j` dependency, also include `-Dlog4j.configuration=file:conf/log4j.properties`

pluradj commented on gremlin-archetype/gremlin-archetype-server/src/main/resources/archetype-resources/src/main/resources/gremlin-server.yaml:
```diff
@@ -0,0 +1,45 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
since `gremlin-server` is a `test` scope dependency, this should move over to `.../src/test/resources` in the generated project

pluradj commented:
I tested both (server and tinkergraph) generated projects on Mac and Windows.
Good stuff @spmallette 

VOTE +1

spmallette commented:
@pluradj that was a nice body of feedback - thanks.  i've cleaned up and added your suggestions.  deploying a fresh SNAPSHOT now.

@okram there really isn't any documentation to "re-publish". i only made one change to our generated asciidoc which added a note about "archetypes" to the reference docs.  the other asciidoc i have in this PR is just README's that get generated into a example project when the archetype is used. 

I could see where a tutorial could be built around an archetype. In the more complex thoughts you had for "examples" I think that would be the way to do it.  Write a tutorial in our standard pattern then build an archetype to support it.
